Sources/SwiftProtobuf/Google_Protobuf_Duration+Extensions.swift

```diff
@@ -20,191 +20,215 @@ private let minDurationSeconds: Int64 = -maxDurationSeconds
-    } else {
-        return String(format: "%ld.%09ds", seconds, abs(nanos))
-    }
+  let (seconds, nanos) = normalizeForDuration(seconds: seconds, nanos: nanos)
+  guard seconds >= minDurationSeconds && seconds <= maxDurationSeconds else {
+    return nil
+  }
+
+  if nanos == 0 {
+    return String(format: "%lds", seconds)
+  } else if nanos % 1000000 == 0 {
+    return String(format: "%ld.%03ds", seconds, abs(nanos) / 1000000)
+  } else if nanos % 1000 == 0 {
+    return String(format: "%ld.%06ds", seconds, abs(nanos) / 1000)
+  } else {
+    return String(format: "%ld.%09ds", seconds, abs(nanos))
+  }
 }
 
 public extension Google_Protobuf_Duration {
-    public init(seconds: Int64 = 0, nanos: Int32 = 0) {
-        self.init()
-        self.seconds = seconds
-        self.nanos = nanos
-    }
+  /// Creates a new `Google_Protobuf_Duration` equal to the given number of
+  /// seconds and nanoseconds.
+  ///
+  /// - Parameter seconds: The number of seconds.
+  /// - Parameter nanos: The number of nanoseconds.
+  public init(seconds: Int64 = 0, nanos: Int32 = 0) {
+    self.init()
+    self.seconds = seconds
+    self.nanos = nanos
+  }
 }
 
 extension Google_Protobuf_Duration: _CustomJSONCodable {
-    mutating func decodeJSON(from decoder: inout JSONDecoder) throws {
-        let s = try decoder.scanner.nextQuotedString()
-        (seconds, nanos) = try parseDuration(text: s)
-    }
-    func encodedJSONString() throws -> String {
-        if let formatted = formatDuration(seconds: seconds, nanos: nanos) {
-            return "\"\(formatted)\""
-        } else {
-            throw JSONEncodingError.durationRange
-        }
+  mutating func decodeJSON(from decoder: inout JSONDecoder) throws {
+    let s = try decoder.scanner.nextQuotedString()
+    (seconds, nanos) = try parseDuration(text: s)
+  }
+  func encodedJSONString() throws -> String {
+    if let formatted = formatDuration(seconds: seconds, nanos: nanos) {
+      return "\"\(formatted)\""
+    } else {
+      throw JSONEncodingError.durationRange
     }
+  }
 }
 
 extension Google_Protobuf_Duration: ExpressibleByFloatLiteral {
-    public typealias FloatLiteralType = Double
-
-    /// Returns a `Google_Protobuf_Duration` initialized to have the same
-    /// meaning as the given literal, when interpreted as a duration in seconds.
-    /// The result will be rounded to the nearest nano if the input has
-    /// precision beyond 1e-9.
-    public init(floatLiteral value: Double) {
-        let sd = trunc(value)
-        let nd = round((value - sd) * TimeInterval(nanosPerSecond))
-        let (s, n) = normalizeForDuration(seconds: Int64(sd), nanos: Int32(nd))
-        self.init(seconds: s, nanos: n)
-    }
+  public typealias FloatLiteralType = Double
+
+  /// Creates a new `Google_Protobuf_Duration` from a floating point literal
+  /// that is interpreted as a duration in seconds, rounded to the nearest
+  /// nanosecond.
+  public init(floatLiteral value: Double) {
```

Comment:
I started to, but then I noticed in other files I mostly didn't document the parameters of `ExpressibleBy*Literal` initializers. In a sense it's unnecessary since you would never call the initializer directly to begin with. I waffle back and forth on it.